File: idea-plugin/src/main/java/com/zzf/rikki/idea/agent/compat/InMemoryPendingApprovalService.java
package com.zzf.rikki.idea.agent.compat;

import com.zzf.rikki.runtime.port.PendingApprovalPort;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicReference;

public class InMemoryPendingApprovalService implements PendingApprovalPort {
    private record PendingState(PendingCommandRecord record, CompletableFuture<ToolExecutionResult> resultFuture) {
    }

    private final Map<String, PendingState> pendingStates = new ConcurrentHashMap<>();
    private final AtomicReference<String> latestPendingCommandId = new AtomicReference<>();
    private final Map<String, AtomicBoolean> skipFlags = new ConcurrentHashMap<>();
    private final Map<String, Set<String>> whitelistedFamilies = new ConcurrentHashMap<>();
    private final Map<String, Set<String>> alwaysAllowFamilies = new ConcurrentHashMap<>();

    @Override
    public PendingCommandRegistration registerPendingCommand(PendingCommandRecord pendingCommand) {
        String autoDecision = autoDecisionFor(pendingCommand);
        if (autoDecision != null) {
            ToolExecutionResult result = executePending(pendingCommand, autoDecision);
            return new PendingCommandRegistration(pendingCommand, false, result);
        }
        pendingStates.put(
                pendingCommand.getId(),
                new PendingState(pendingCommand, new CompletableFuture<>())
        );
        latestPendingCommandId.set(pendingCommand.getId());
        return new PendingCommandRegistration(pendingCommand, true, null);
    }

    @Override
    public ToolExecutionResult awaitDecision(String commandId) {
        PendingState state = pendingStates.get(commandId);
        if (state == null) {
            return new ToolExecutionResult("rejected", "Pending command not found: " + commandId, "Pending command not found: " + commandId, Map.of(), null, null, null, null, false);
        }
        try {
            return state.resultFuture().get(1L, TimeUnit.HOURS);
        } catch (Exception e) {
            pendingStates.remove(commandId);
            return new ToolExecutionResult("error", "", e.getMessage() == null ? "Timed out waiting for approval." : e.getMessage(), Map.of(), null, null, null, null, false);
        }
    }

    @Override
    public PendingCommandResolutionResult resolve(String commandId, boolean reject, String decisionMode) {
        PendingState state = pendingStates.get(commandId);
        if (state == null) {
            return new PendingCommandResolutionResult("error", "", "Pending command not found: " + commandId, null, "", false, PendingApprovalService.DECISION_MANUAL);
        }
        String normalizedDecision = normalizeDecisionMode(decisionMode);
        ToolExecutionResult result;
        if (reject) {
            LinkedHashMap<String, Object> meta = new LinkedHashMap<>();
            meta.put("approval_result", "rejected");
            meta.put("resolved_by_user", Boolean.TRUE);
            meta.put("approval_mode", PendingApprovalService.DECISION_MANUAL);
            result = new ToolExecutionResult(
                    "rejected",
                    "User rejected command, not executed: " + state.record().getCommand(),
                    null,
                    meta,
                    null,
                    null,
                    null,
                    null,
                    false
            );
        } else {
            applyDecisionPolicy(state.record(), normalizedDecision);
            result = executePending(state.record(), normalizedDecision);
        }
        pendingStates.remove(commandId);
        state.resultFuture().complete(result);
        return result.toResolutionResult(state.record().getCommand(), normalizedDecision);
    }

    @Override
    public void skipCurrentExecution() {
        skipFlags.values().forEach(flag -> flag.set(true));
    }

    @Override
    public void clearSession(String sessionId) {
        if (sessionId == null || sessionId.isBlank()) {
            return;
        }
        pendingStates.values().stream()
                .filter(state -> sessionId.equals(state.record().getSessionId()))
                .map(state -> state.record().getId())
                .toList()
                .forEach(id -> {
                    PendingState state = pendingStates.remove(id);
                    if (state != null) {
                        state.resultFuture().complete(
                                new ToolExecutionResult(
                                        "rejected",
                                        "Pending command cleared for session: " + sessionId,
                                        "Pending command cleared for session: " + sessionId,
                                        Map.of(),
                                        null,
                                        null,
                                        null,
                                        null,
                                        false
                                )
                        );
                    }
                });
        skipFlags.remove(sessionId);
        whitelistedFamilies.remove(sessionId);
        alwaysAllowFamilies.remove(sessionId);
    }

    public PendingCommandResolutionResult resolveLatestPending(boolean approve) {
        String commandId = latestPendingCommandId.get();
        if (commandId == null || commandId.isBlank()) {
            return new PendingCommandResolutionResult("error", "", "No pending command.", null, "", false, PendingApprovalService.DECISION_MANUAL);
        }
        return resolve(commandId, !approve, PendingApprovalService.DECISION_MANUAL);
    }

    public AtomicBoolean skipFlagFor(String sessionId) {
        String key = sessionId == null || sessionId.isBlank() ? "__default__" : sessionId;
        return skipFlags.computeIfAbsent(key, ignored -> new AtomicBoolean(false));
    }

    public static String normalizeDecisionMode(String decisionMode) {
        if (PendingApprovalService.DECISION_WHITELIST.equalsIgnoreCase(decisionMode)) {
            return PendingApprovalService.DECISION_WHITELIST;
        }
        if (PendingApprovalService.DECISION_ALWAYS_ALLOW_NON_DESTRUCTIVE.equalsIgnoreCase(decisionMode)) {
            return PendingApprovalService.DECISION_ALWAYS_ALLOW_NON_DESTRUCTIVE;
        }
        return PendingApprovalService.DECISION_MANUAL;
    }

    private String autoDecisionFor(PendingCommandRecord record) {
        if (record.getSessionId() == null || record.getSessionId().isBlank() || record.getStrictApproval()) {
            return null;
        }
        if (alwaysAllowFamilies.getOrDefault(record.getSessionId(), Set.of()).contains(record.getCommandFamily())) {
            return PendingApprovalService.DECISION_ALWAYS_ALLOW_NON_DESTRUCTIVE;
        }
        if (whitelistedFamilies.getOrDefault(record.getSessionId(), Set.of()).contains(record.getCommandFamily())) {
            return PendingApprovalService.DECISION_WHITELIST;
        }
        return null;
    }

    private void applyDecisionPolicy(PendingCommandRecord record, String decisionMode) {
        if (record.getSessionId() == null || record.getSessionId().isBlank()) {
            return;
        }
        if (PendingApprovalService.DECISION_WHITELIST.equals(decisionMode)) {
            whitelistedFamilies.computeIfAbsent(record.getSessionId(), ignored -> ConcurrentHashMap.newKeySet()).add(record.getCommandFamily());
        }
        if (PendingApprovalService.DECISION_ALWAYS_ALLOW_NON_DESTRUCTIVE.equals(decisionMode) && !record.getStrictApproval()) {
            alwaysAllowFamilies.computeIfAbsent(record.getSessionId(), ignored -> ConcurrentHashMap.newKeySet()).add(record.getCommandFamily());
        }
    }

    private ToolExecutionResult executePending(PendingCommandRecord record, String decisionMode) {
        AtomicBoolean skipFlag = skipFlagFor(record.getSessionId());
        skipFlag.set(false);
        ToolExecutionResult raw;
        try {
            raw = record.getExecutor().get();
        } catch (Exception e) {
            raw = new ToolExecutionResult("error", "", e.getMessage() == null ? "Failed to execute pending command." : e.getMessage(), Map.of(), null, null, null, null, false);
        }
        LinkedHashMap<String, Object> meta = new LinkedHashMap<>(raw.getMeta());
        meta.put("approval_result", raw.getStatus());
        meta.put("resolved_by_user", Boolean.TRUE);
        meta.put("approval_mode", decisionMode);
        return new ToolExecutionResult(
                raw.getStatus(),
                raw.getOutput(),
                raw.getError(),
                meta,
                raw.getPendingChange(),
                raw.getPendingCommand(),
                raw.getTodoJson(),
                raw.getExitCode(),
                raw.getTimeout()
        );
    }
}

File: idea-plugin/src/main/kotlin/com/zzf/rikki/idea/agent/compat/InMemoryPendingApprovalService.kt
package com.zzf.rikki.idea.agent.compat

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicReference

class InMemoryPendingApprovalService : PendingApprovalService {
    private data class PendingState(
        val record: PendingCommandRecord,
        val resultFuture: CompletableFuture<ToolExecutionResult>
    )

    private val pendingStates = ConcurrentHashMap<String, PendingState>()
    private val latestPendingCommandId = AtomicReference<String?>()
    private val skipFlags = ConcurrentHashMap<String, AtomicBoolean>()
    private val whitelistedFamilies = ConcurrentHashMap<String, MutableSet<String>>()
    private val alwaysAllowFamilies = ConcurrentHashMap<String, MutableSet<String>>()

    override fun registerPendingCommand(pendingCommand: PendingCommandRecord): PendingCommandRegistration {
        val autoDecision = autoDecisionFor(pendingCommand)
        if (autoDecision != null) {
            val result = executePending(pendingCommand, autoDecision)
            return PendingCommandRegistration(
                pendingCommand = pendingCommand,
                awaitApproval = false,
                immediateResult = result
            )
        }
        pendingStates[pendingCommand.id] = PendingState(
            record = pendingCommand,
            resultFuture = CompletableFuture()
        )
        latestPendingCommandId.set(pendingCommand.id)
        return PendingCommandRegistration(
            pendingCommand = pendingCommand,
            awaitApproval = true
        )
    }

    override suspend fun awaitDecision(commandId: String): ToolExecutionResult = withContext(Dispatchers.IO) {
        val state = pendingStates[commandId]
        if (state == null) {
            ToolExecutionResult(
                status = "rejected",
                output = "Pending command not found: $commandId",
                error = "Pending command not found: $commandId"
            )
        } else {
            try {
                state.resultFuture.get(1L, TimeUnit.HOURS)
            } catch (e: Exception) {
                pendingStates.remove(commandId)
                ToolExecutionResult(
                    status = "error",
                    output = "",
                    error = e.message ?: "Timed out waiting for approval."
                )
            }
        }
    }

    override fun resolve(
        commandId: String,
        reject: Boolean,
        decisionMode: String
    ): PendingCommandResolutionResult {
        val state = pendingStates[commandId]
            ?: return PendingCommandResolutionResult(
                status = "error",
                output = "",
                error = "Pending command not found: $commandId"
            )
        val normalizedDecision = normalizeDecisionMode(decisionMode)
        val result = if (reject) {
            ToolExecutionResult(
                status = "rejected",
                output = "User rejected command, not executed: ${state.record.command}",
                meta = mapOf(
                    "approval_result" to "rejected",
                    "resolved_by_user" to true,
                    "approval_mode" to PendingApprovalService.DECISION_MANUAL
                )
            )
        } else {
            applyDecisionPolicy(state.record, normalizedDecision)
            executePending(state.record, normalizedDecision)
        }
        pendingStates.remove(commandId)
        state.resultFuture.complete(result)
        return result.toResolutionResult(state.record.command, normalizedDecision)
    }

    override fun skipCurrentExecution() {
        skipFlags.values.forEach { it.set(true) }
    }

    override fun clearSession(sessionId: String) {
        if (sessionId.isBlank()) {
            return
        }
        val ids = pendingStates.values
            .filter { it.record.sessionId == sessionId }
            .map { it.record.id }
        for (id in ids) {
            val state = pendingStates.remove(id) ?: continue
            state.resultFuture.complete(
                ToolExecutionResult(
                    status = "rejected",
                    output = "Pending command cleared for session: $sessionId",
                    error = "Pending command cleared for session: $sessionId"
                )
            )
        }
        skipFlags.remove(sessionId)
        whitelistedFamilies.remove(sessionId)
        alwaysAllowFamilies.remove(sessionId)
    }

    fun resolveLatestPending(approve: Boolean): PendingCommandResolutionResult {
        val commandId = latestPendingCommandId.get()
            ?: return PendingCommandResolutionResult(
                status = "error",
                output = "",
                error = "No pending command."
            )
        return resolve(
            commandId = commandId,
            reject = !approve,
            decisionMode = PendingApprovalService.DECISION_MANUAL
        )
    }

    fun skipFlagFor(sessionId: String): AtomicBoolean =
        skipFlags.computeIfAbsent(sessionId.ifBlank { "__default__" }) { AtomicBoolean(false) }

    private fun autoDecisionFor(record: PendingCommandRecord): String? {
        if (record.sessionId.isBlank() || record.strictApproval) {
            return null
        }
        if (alwaysAllowFamilies[record.sessionId]?.contains(record.commandFamily) == true) {
            return PendingApprovalService.DECISION_ALWAYS_ALLOW_NON_DESTRUCTIVE
        }
        if (whitelistedFamilies[record.sessionId]?.contains(record.commandFamily) == true) {
            return PendingApprovalService.DECISION_WHITELIST
        }
        return null
    }

    private fun applyDecisionPolicy(record: PendingCommandRecord, decisionMode: String) {
        if (record.sessionId.isBlank()) {
            return
        }
        if (decisionMode == PendingApprovalService.DECISION_WHITELIST) {
            val set = whitelistedFamilies.computeIfAbsent(record.sessionId) {
                ConcurrentHashMap.newKeySet<String>()
            }
            set += record.commandFamily
        }
        if (
            decisionMode == PendingApprovalService.DECISION_ALWAYS_ALLOW_NON_DESTRUCTIVE
            && !record.strictApproval
        ) {
            val set = alwaysAllowFamilies.computeIfAbsent(record.sessionId) {
                ConcurrentHashMap.newKeySet<String>()
            }
            set += record.commandFamily
        }
    }

    private fun executePending(
        record: PendingCommandRecord,
        decisionMode: String
    ): ToolExecutionResult {
        val skipFlag = skipFlagFor(record.sessionId)
        skipFlag.set(false)
        val raw = try {
            record.executor()
        } catch (e: Exception) {
            ToolExecutionResult(
                status = "error",
                output = "",
                error = e.message ?: "Failed to execute pending command."
            )
        }
        val mergedMeta = LinkedHashMap<String, Any?>()
        mergedMeta.putAll(raw.meta)
        mergedMeta["approval_result"] = raw.status
        mergedMeta["resolved_by_user"] = true
        mergedMeta["approval_mode"] = decisionMode
        return raw.copy(meta = mergedMeta)
    }

    companion object {
        fun normalizeDecisionMode(decisionMode: String?): String = when (decisionMode?.trim()?.lowercase()) {
            PendingApprovalService.DECISION_WHITELIST -> PendingApprovalService.DECISION_WHITELIST
            PendingApprovalService.DECISION_ALWAYS_ALLOW_NON_DESTRUCTIVE ->
                PendingApprovalService.DECISION_ALWAYS_ALLOW_NON_DESTRUCTIVE

            else -> PendingApprovalService.DECISION_MANUAL
        }
    }
}